doc/man/bitcoind.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH BITCOIND "1" "March 2026" "bitcoind v1.2.1.0" "User Commands"
.SH NAME
bitcoind \- manual page for bitcoind v1.2.1.0
.SH DESCRIPTION
Bitcoin SV version v1.2.1.0\-dev
.SS "Usage:"
.TP
bitcoind [options]
Start Bitcoin SV Daemon
.SH OPTIONS
.HP
\-?
.IP
Print this help message and exit
.HP
\fB\-version\fR
.IP
Print version and exit
.HP
\fB\-alertnotify=\fR<cmd>
.IP
Execute command when a relevant alert is received or we see a really
long fork (%s in cmd is replaced by message)
.HP
\fB\-blocknotify=\fR<cmd>
.IP
Execute command when the best block changes (%s in cmd is replaced by
block hash)
.HP
\fB\-assumevalid=\fR<hex>
.IP
If this block is in the chain assume that it and its ancestors are valid
and potentially skip their script verification (0 to verify all,
default:
000000000000000000e45ad2fbcc5ff3e85f0868dd8f00ad4e92dffabe28f8d2,
testnet:
0000000000327972b8470c11755adf8f4319796bafae01f5a6650490b98a17db)
.HP
\fB\-conf=\fR<file>
.IP
Specify configuration file (default: bitcoin.conf)
.HP
\fB\-daemon\fR
.IP
Run in the background as a daemon and accept commands
.HP
\fB\-datadir=\fR<dir>
.IP
Specify data directory
.HP
\fB\-dbcache=\fR<n>
.IP
Set database cache size in megabytes (4 to 16384, default: 2048). The
value may be given in megabytes or with unit (B, KiB, MiB, GiB).
.HP
\fB\-frozentxodbcache=\fR<n>
.IP
Set cache size for database holding a list of frozen transaction outputs
in bytes (default: 128000)
.HP
\fB\-genesisactivationheight\fR
.IP
Set block height at which genesis should be activated. (default:
620538).
.HP
\fB\-loadblock=\fR<file>
.IP
Imports blocks from external blk000??.dat file on startup
.HP
\fB\-maxmempool=\fR<n>
.IP
Keep the resident size of the transaction memory pool below <n>
megabytes (default: 1000,  must be at least 300). The value may
be given in megabytes or with unit (B, kB, MB, GB).
.HP
\fB\-mempoolmaxpercentcpfp=\fR<n>
.IP
Percentage of total mempool size (ram+disk) to allow for low paying
transactions (0..100) (default: 10)
.HP
\fB\-mempoolexpiry=\fR<n>
.IP
Do not keep transactions in the mempool longer than <n> hours (default:
336)
.HP
\fB\-maxmempoolnonfinal=\fR<n>
.IP
Keep the non\-final transaction memory pool below <n> megabytes (default:
50). The value may be given in megabytes or with unit (B, KiB,
MiB, GiB).
.HP
\fB\-mempoolexpirynonfinal=\fR<n>
.IP
Do not keep transactions in the non\-final mempool longer than <n> hours
(default: 672)
.HP
\fB\-mempoolnonfinalmaxreplacementrate=\fR<n>
.IP
The maximum rate at which a transaction in the non\-final mempool can be
replaced by another updated transaction, expressed as
transactions per hour. (default: 7200/hour)
.HP
\fB\-persistmempool\fR
.IP
Whether to save the mempool on shutdown and load on restart (default: 1)
.HP
\fB\-threadsperblock=\fR<n>
.IP
Set the number of script verification threads used when validating
single block (0 to 64, 0 = auto, default: 0)
.HP
\fB\-txnthreadsperblock=\fR<n>
.IP
Set the number of transaction verification threads used when validating
single block (0 to 64, 0 = auto, default: 0)
.HP
\fB\-scriptvalidatormaxbatchsize=\fR<n>
.IP
Set size of script verification batch per thread (1 to 255, default:
128)
.HP
\fB\-maxparallelblocks=\fR<n>
.IP
Set the number of block that can be validated in parallel across all
nodes. If additional block arrive, validation of an old block is
terminated. (1 to 100, default: 4)
.HP
\fB\-maxparallelblocksperpeer=\fR<n>
.IP
Set the number of blocks that can be validated in parallel from a single
peer. If peers sends another block, the validation of it is
delayed. (1 to maxparallelblocks, default: 3)
.HP
\fB\-pid=\fR<file>
.IP
Specify pid file (default: bitcoind.pid)
.HP
\fB\-preload=\fR<n>
.IP
If n is set to 1, blockchain state will be preloaded into memory. If n
is 0, no preload will happen. Other values for n are not allowed.
The default value is 0. This option is not supported on Windows
operating systems.
.HP
\fB\-prune=\fR<n>
.IP
Reduce storage requirements by enabling pruning (deleting) of old
blocks. This allows the pruneblockchain RPC to be called to
delete specific blocks, and enables automatic pruning of old
blocks if a target size in MiB is provided. This mode is
incompatible with \fB\-txindex\fR and \fB\-rescan\fR. Warning: Reverting this
setting requires re\-downloading the entire blockchain. (default:
0 = disable pruning blocks, 1 = allow manual pruning via RPC,
>550 = automatically prune block files to stay under the
specified target size in MiB, but still keep the last 288 blocks
to speed up a potential reorg even if this results in the pruning
target being exceeded)Note: Currently achievable prune target is
~100GB (mainnet). Setting the target size too low will not affect
pruning function, but will not guarantee block files size staying
under the threshold at all times.
.HP
\fB\-reindex\-chainstate\fR
.IP
Rebuild chain state from the currently indexed blocks
.HP
\fB\-reindex\fR
.IP
Rebuild chain state and block index from the blk*.dat files on disk
.HP
\fB\-rejectmempoolrequest\fR
.IP
Reject every mempool request from non\-whitelisted peers (default: 1).
.HP
\fB\-sysperms\fR
.IP
Create new files with system default permissions, instead of umask 077
(only effective with disabled wallet functionality)
.HP
\fB\-txindex\fR
.IP
Maintain a full transaction index, used by the getrawtransaction rpc
call (default: 0)
.HP
\fB\-preferredblockfilesize\fR
.IP
Preferred size (in bytes) of a single datafile containing blocks. When
this size is reached, new datafile is created. If preferred size
is less than size of a single block, block will still be stored,
meaning datafile size can be larger than preferred size.
(default: 134217728).
.HP
\fB\-maxmerkletreediskspace\fR
.IP
Maximum disk size in bytes that can be taken by stored merkle trees.
This size should not be less than default size (default: 921MB
for a maximum 4GB block size). The value may be given in bytes or
with unit (B, kiB, MiB, GiB).
.HP
\fB\-preferredmerkletreefilesize\fR
.IP
Preferred size of a single datafile containing merkle trees. When size
is reached, new datafile is created. If preferred size is less
than size of a single merkle tree, it will still be stored,
meaning datafile size can be larger than preferred size.
(default: 64MB for a maximum 4GB block size). The value may be
given in bytes or with unit (B, kiB, MiB, GiB).
.HP
\fB\-maxmerkletreememcachesize\fR
.IP
Maximum merkle trees memory cache size in bytes. For faster responses,
requested merkle trees are stored into a memory cache. (default:
64MB for a maximum 4GB block size). The value may be given in
bytes or with unit (B, kiB, MiB, GiB).
.PP
Connection options:
.HP
\fB\-addnode=\fR<ip>
.IP
Add a node to connect to and attempt to keep the connection open
.HP
\fB\-banscore=\fR<n>
.IP
Threshold for disconnecting misbehaving peers (default: 100)
.HP
\fB\-bantime=\fR<n>
.IP
Number of seconds to keep misbehaving peers from reconnecting (default:
86400)
.HP
\fB\-bind=\fR<addr>
.IP
Bind to given address and always listen on it. Use [host]:port notation
for IPv6
.HP
\fB\-blockstallingmindownloadspeed=\fR<n>
.IP
Minimum average download speed (Kbytes/s) we will allow a stalling peer
to fall to during IBD. A value of 0 means stall detection is
disabled (default: 100Kb/s)
.HP
\fB\-broadcastdelay=\fR<n>
.IP
Set inventory broadcast delay duration in millisecond(min: 0, max:
50000)
.HP
\fB\-connect=\fR<ip>
.IP
Connect only to the specified node(s); \fB\-noconnect\fR or \fB\-connect\fR=\fI\,0\/\fR alone to
disable automatic connections
.HP
\fB\-discover\fR
.IP
Discover own IP addresses (default: 1 when listening and no \fB\-externalip\fR
or \fB\-proxy\fR)
.HP
\fB\-dns\fR
.IP
Allow DNS lookups for \fB\-addnode\fR, \fB\-seednode\fR and \fB\-connect\fR (default: 1)
.HP
\fB\-dnsseed\fR
.IP
Query for peer addresses via DNS lookup, if low on addresses (default: 1
unless \fB\-connect\fR/\-noconnect)
.HP
\fB\-externalip=\fR<ip>
.IP
Specify your own public address
.HP
\fB\-forcednsseed\fR
.IP
Always query for peer addresses via DNS lookup (default: 1)
.HP
\fB\-listen\fR
.IP
Accept connections from outside (default: 1 if no \fB\-proxy\fR or
\fB\-connect\fR/\-noconnect)
.HP
\fB\-maxaddnodeconnections=\fR<n>
.IP
Maximum number of additional outgoing connections to maintain that have
been added via addnode (default: 8)
.HP
\fB\-maxblocktxnpercent=\fR<n>
.IP
Maximum perentage of txns from a block we will respond to a getblocktxn
request with a blocktxn response. Larger than this we will just
respond with the entire block (default: 99)
.HP
\fB\-maxoutboundconnections=\fR<n>
.IP
Maintain at most <n> outbound connections to peers (default: 8)
.HP
\fB\-maxconnectionsfromaddr=\fR<n>
.IP
Maximum number of inbound connections from a single address (not
applicable to whitelisted peers) 0 = unrestricted (default: 0)
.HP
\fB\-maxconnections=\fR<n>
.IP
Maintain at most <n> connections to peers (default: 125)
.HP
\fB\-maxreceivebuffer=\fR<n>
.IP
Maximum per\-connection receive buffer in kilobytes (default: 500000).
The value may be given in kilobytes or with unit (B, kB, MB, GB).
.HP
\fB\-maxsendbuffer=\fR<n>
.IP
Maximum per\-connection send buffer in kilobytes (default: 500000). The
value may be given in kilobytes or with unit (B, kB, MB, GB).
.HP
\fB\-factormaxsendqueuesbytes=\fR<n>
.IP
Factor that will be multiplied with excessiveBlockSize to limit the
maximum bytes in all sending queues. If this size is exceeded, no
response to block related P2P messages is sent. (default factor:
4)
.HP
\fB\-maxtimeadjustment\fR
.IP
Maximum allowed median peer time offset adjustment. Local perspective of
time may be influenced by peers forward or backward by this
amount. (default: 4200 seconds)
.HP
\fB\-multistreams\fR
.IP
Enable the use of multiple streams to our peers (default: 1)
.HP
\fB\-multistreampolicies\fR
.IP
List of stream policies to use with our peers in order of preference
(available policies: BlockPriority,Default, default:
BlockPriority,Default)
.HP
\fB\-onlynet=\fR<net>
.IP
Only connect to nodes in network <net> (ipv4 or ipv6)
.HP
\fB\-permitbaremultisig\fR
.IP
Relay non\-P2SH multisig (default: 1)
.HP
\fB\-peerbloomfilters\fR
.IP
Support filtering of blocks and transaction with bloom filters (default:
0)
.HP
\fB\-port=\fR<port>
.IP
Listen for connections on <port> (default: 8333 or testnet: 18333)
.HP
\fB\-proxy=\fR<ip:port>
.IP
Connect through SOCKS5 proxy
.HP
\fB\-proxyrandomize\fR
.IP
Randomize credentials for every proxy connection. (default: 1)
.HP
\fB\-seednode=\fR<ip>
.IP
Connect to a node to retrieve peer addresses, and disconnect
.HP
\fB\-timeout=\fR<n>
.IP
Specify connection timeout in milliseconds (minimum: 1, default: 5000)
.HP
\fB\-upnp\fR
.IP
Use UPnP to map the listening port (default: 0)
.HP
\fB\-whitebind=\fR<addr>
.IP
Bind to given address and whitelist peers connecting to it. Use
[host]:port notation for IPv6
.HP
\fB\-whitelist=\fR<IP address or network>
.IP
Whitelist peers connecting from the given IP address (e.g. 1.2.3.4) or
CIDR notated network (e.g. 1.2.3.0/24). Can be specified multiple
times. Whitelisted peers cannot be DoS banned and their
transactions are always relayed, even if they are already in the
mempool, useful e.g. for a gateway
.HP
\fB\-whitelistrelay\fR
.IP
Accept relayed transactions received from whitelisted peers even when
not relaying transactions (default: 1)
.HP
\fB\-whitelistforcerelay\fR
.IP
Force relay of transactions from whitelisted peers even if they violate
local relay policy (default: 1)
.HP
\fB\-maxuploadtarget=\fR<n>
.IP
Tries to keep outbound traffic under the given target (in MiB per 24h),
0 = no limit (default: 0). The value may be given in megabytes or
with unit (KiB, MiB, GiB).
.HP
\fB\-maxpendingresponses_getheaders=\fR<n>
.IP
Maximum allowed number of pending responses in the sending queue for
received GETHEADERS P2P requests before the connection is closed.
Not applicable to whitelisted peers. 0 = no limit (default: 0).
Main purpose of this setting is to limit memory usage. The
specified value should be small (e.g. ~50) since in practice
connected peers do not need to send many GETHEADERS requests in
parallel.
.HP
\fB\-maxpendingresponses_gethdrsen=\fR<n>
.IP
Maximum allowed number of pending responses in the sending queue for
received GETHDRSEN P2P requests before the connection is closed.
Not applicable to whitelisted peers. 0 = no limit (default: 0).
Main purpose of this setting is to limit memory usage. The
specified value should be small (e.g. ~10) since in practice
connected peers do not need to send many GETHDRSEN requests in
parallel.
.PP
Wallet options:
.HP
\fB\-disablewallet\fR
.IP
Do not load the wallet and disable wallet RPC calls
.HP
\fB\-keypool=\fR<n>
.IP
Set key pool size to <n> (default: 1000)
.HP
\fB\-fallbackfee=\fR<amt>
.IP
A fee rate (in BSV/kB) that will be used when fee estimation has
insufficient data (default: 0.0002)
.HP
\fB\-mintxfee=\fR<amt>
.IP
Fees (in BSV/kB) smaller than this are considered zero fee for
transaction creation (default: 0.00001)
.HP
\fB\-paytxfee=\fR<amt>
.IP
Fee (in BSV/kB) to add to transactions you send (default: 0.00)
.HP
\fB\-rescan\fR
.IP
Rescan the block chain for missing wallet transactions on startup
.HP
\fB\-salvagewallet\fR
.IP
Attempt to recover private keys from a corrupt wallet on startup
.HP
\fB\-spendzeroconfchange\fR
.IP
Spend unconfirmed change when sending transactions (default: 1)
.HP
\fB\-usehd\fR
.IP
Use hierarchical deterministic key generation (HD) after BIP32. Only has
effect during wallet creation/first start (default: 1)
.HP
\fB\-upgradewallet\fR
.IP
Upgrade wallet to latest format on startup
.HP
\fB\-wallet=\fR<file>
.IP
Specify wallet file (within data directory) (default: wallet.dat)
.HP
\fB\-walletbroadcast\fR
.IP
Make the wallet broadcast transactions (default: 1)
.HP
\fB\-walletnotify=\fR<cmd>
.IP
Execute command when a wallet transaction changes (%s in cmd is replaced
by TxID)
.HP
\fB\-zapwallettxes=\fR<mode>
.IP
Delete all wallet transactions and only recover those parts of the
blockchain through \fB\-rescan\fR on startup (1 = keep tx meta data e.g.
account owner and payment request information, 2 = drop tx meta
data)
.PP
ZeroMQ notification options:
.HP
\fB\-zmqpubhashblock=\fR<address>
.IP
Enable publish hash block in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubhashtx=\fR<address>
.IP
Enable publish hash transaction in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubrawblock=\fR<address>
.IP
Enable publish raw block in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubrawtx=\fR<address>
.IP
Enable publish raw transaction in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubinvalidtx=\fR<address>
.IP
Enable publish invalid transaction in <address>. \fB\-invalidtxsink\fR=\fI\,ZMQ\/\fR
should be specified. For more information see doc/zmq.md.
.HP
\fB\-zmqpubremovedfrommempool=\fR<address>
.IP
Enable publish removal of transaction (txid and the reason in json
format) in <address>. For more information see doc/zmq.md.
.HP
\fB\-zmqpubremovedfrommempoolblock=\fR<address>
.IP
Enable publish removal of transaction (txid and the reason in json
format) in <address>. For more information see doc/zmq.md.
.HP
\fB\-zmqpubhashtx2=\fR<address>
.IP
Enable publish hash transaction in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubrawtx2=\fR<address>
.IP
Enable publish raw transaction in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubhashblock2=\fR<address>
.IP
Enable publish hash block in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubrawblock2=\fR<address>
.IP
Enable publish raw block in <address>. For more information see
doc/zmq.md.
.PP
Debugging/Testing options:
.HP
\fB\-uacomment=\fR<cmt>
.IP
Append comment to the user agent string
.HP
\fB\-debug=\fR<category>
.IP
Output debugging information (default: 0, supplying <category> is
optional). If <category> is not supplied or if <category> = 1,
output all debugging information.<category> can be: mempool,
http, bench, zmq, db, rpc, addrman, selectcoins, reindex,
cmpctblock, rand, prune, proxy, mempoolrej, libevent, coindb,
leveldb, txnprop, txnsrc, journal, txnval, netconn, netmsg,
netmsgverb, netmsgall, net, doublespend, minerid, blocksrc.
.HP
\fB\-debugexclude=\fR<category>
.IP
Exclude debugging information for a category. Can be used in conjunction
with \fB\-debug\fR=\fI\,1\/\fR to output debug logs for all categories except one
or more specified categories.
.HP
\fB\-help\-debug\fR
.IP
Show all debugging options (usage: \fB\-\-help\fR \fB\-help\-debug\fR)
.HP
\fB\-debugp2pthreadstalls\fR
.IP
Log P2P requests that stall request processing loop for longer than
specified milliseconds (default: disabled)
.HP
\fB\-logips\fR
.IP
Include IP addresses in debug output (default: 0)
.HP
\fB\-logtimestamps\fR
.IP
Prepend debug output with timestamp (default: 1)
.HP
\fB\-maxtxfee=\fR<amt>
.IP
Maximum total fees (in BSV) to use in a single wallet transaction or raw
transaction; setting this too low may abort large transactions
(default: 0.10)
.HP
\fB\-printtoconsole\fR
.IP
Send trace/debug info to console instead of bitcoind.log file
.HP
\fB\-shrinkdebugfile\fR
.IP
Shrink bitcoind.log file on client startup (default: 1 when no \fB\-debug\fR)
.PP
Chain selection options:
.HP
\fB\-testnet\fR
.IP
Use the test chain
.HP
\fB\-regtest\fR
.IP
Enter regression test mode, which uses a special chain in which blocks
can be solved instantly. This is intended for regression testing
tools and app development.
.HP
\fB\-stn\fR
.IP
Use the Scaling Test Network
.PP
Node relay options:
.HP
\fB\-excessiveblocksize=\fR<n>
.IP
Set the maximum block size in bytes we will accept from any source. This
is the effective block size hard limit and it is a required
parameter (0 = unlimited). The value may be given in bytes or
with unit (B, kB, MB, GB).
.HP
\fB\-datacarrier\fR
.IP
Relay and mine data carrier transactions (default: 1)
.HP
\fB\-datacarriersize\fR
.IP
Maximum size of data in data carrier transactions we relay and mine
(default: 4294967295). The value may be given in bytes or with
unit (B, kB, MB, GB).
.HP
\fB\-maxstackmemoryusageconsensus\fR
.IP
Set maximum stack memory usage in bytes used for script verification
we're willing to accept from any source (0 = unlimited) after
Genesis is activated (consensus level). This is a required
parameter. The value may be given in bytes or with unit (B, kB,
MB, GB).
.HP
\fB\-maxstackmemoryusagepolicy\fR
.IP
Set maximum stack memory usage used for script verification we're
willing to relay/mine in a single transaction (default: 100 MB, 0
= unlimited) after Genesis is activated (policy level). The value
may be given in bytes or with unit (B, kB, MB, GB). Must be less
or equal to \fB\-maxstackmemoryusageconsensus\fR.
.HP
\fB\-maxopsperscriptpolicy=\fR<n>
.IP
Set maximum number of non\-push operations we're willing to relay/mine
per script (default: unlimited, 0 = unlimited), after Genesis is
activated
.HP
\fB\-maxtxsigopscountspolicy=\fR<n>
.IP
Set maximum allowed number of signature operations we're willing to
relay/mine in a single transaction (default: unlimited, 0 =
unlimited) after Genesis is activated.
.HP
\fB\-maxstdtxvalidationduration=\fR<n>
.IP
Set the single standard transaction validation duration threshold in
milliseconds after which the standard transaction validation will

doc/release-notes.md
# Bitcoin SV Node Software – v1.2.0 Release

This release is a hard fork which includes changes to the BSV protocol.   

The scheduled TestNet activation height is 1,686,611 (Target is 12:00 midday 05-Aug-2025)

The scheduled MainNet activation height is 921,788 (Target is 12:00 midday 04-Nov-2025)

## What’s Changed  

### Chronicle Protocol Updates

#### Script OpCodes

The following opcodes are restored:  

*   **OP\_VER** – pushes the transaction version (4 byte array) onto the top of the stack.
*   **OP\_VERIF, OP\_VERNOTIF** – provides conditional logic based on the transaction version.
*   **OP\_SUBSTR** – creates an arbitrary substring of the string on top of the stack.
*   **OP\_LEFT, OP\_RIGHT –** creates the left/right most substring of specific length from the string on the top of the stack respectively.
*   **OP\_2MUL** – doubles the number on top of the stack
*   **OP\_2DIV** – halves the number on top of the stack

Please refer to the Protocol Restoration specification for further details.  

#### Sighash Flag Changes

A new optional sighash flag **SIGHASH_CHRONICLE** is introduced (value **0x20**).

Transactions signed without the CHRONICLE flag will see no changes to their
consensus rules and will remain largely unaffected by the Chronicle update.
Transactions signed *with* the CHRONICLE flag will see the changes described
in the following sections.

#### Transaction Digest Algorithm

The BSV blockchain will again support the original Bitcoin transaction digest
algorithm for transactions signed with the CHRONICLE sighash flag. Transactions
signed without the CHRONICLE flag will continue to use the BIP143 new transaction
digest algorithm as before.

#### Transaction Limitations Relaxed

The limitations in the following areas are removed.   

*   **Script numbers** – Limits on the maximum size of script numbers are removed.
Please note that practical limits are imposed by the external libraries used to
implement script numbers (max size is currently 64MB).
*   **Minimal Encoding Requirement** – Previously releases required transactions
to encode numbers as efficiently as possible. For example, the number two can be
encoded as 0002 (2 bytes) or 02 (1 byte). Prior to this release only the second
version is accepted. Minimal encoding places an unnecessary burden on users that
was not present in the original Satoshi implementation and it is removed from this
release for transactions signed with the CHRONICLE sighash flag.
*   **Low S signatures** - If S is a signature, then so is -S. The “low S” signature
requirement was introduced to decrease transaction malleability but is an unnecessary
requirement and is removed. Either signature S or -S can now be used in transactions
signed with the CHRONICLE sighash flag.
*   **NULLFAIL checks** - The requirement that if an `OP_CHECKSIG` or
`OP_CHECKMULTISIG` is trying to return a `FALSE` value to the stack then signatures
must be an empty array is removed for transactions signed with the CHRONICLE sighash
flag.
*   **NULLDUMMY checks** - The requirement that the dummy value in the unlocking
script for a multisig transaction must be an empty array is removed for transactions
signed with the CHRONICLE sighash flag.
*   **Clean Stack Policy** – Previous releases required that after the execution
of the unlocking and locking script, the stack is “clean”. i.e. there is only a
single item (interpreted as true/false) on the stack that indicates whether the
transaction has permission to spend the relevant input. The requirement is
unnecessary and adds complexity to scripts. This release removes the clean stack
requirement for transactions signed with the CHRONICLE sighash flag.
*   **No opcodes in Unlocking Scripts Policy** - Previous releases do not allow
unlocking scripts to include non-data opcodes. That restriction is removed in
this release for transactions signed with the CHRONICLE sighash flag.

### Performance Improvements

#### Memory Usage

TCMalloc is now the default memory allocator used when building bitcoind. In
testing, this change results in peak memory usage of the node dropping by nearly
a half.

#### Initial Block Download (IBD)

By tuning some existing leveldb configuration parameters, introducing some new
configuration options and optimising the code, the time taken to perform a
complete IBD has been reduced. See the section below for details on the new
config options.

### Config Variables

#### Mempool Synchonisation

A new feature has been introduced to allow nodes to be configured to periodically
update their mempools with missing transactions from other specified peers. Note
that peers that wish to synchronise mempools must be configured symmetrically, i.e.
both peers must be configured to sync' with each other.

The options to configure this are:

*   **mempoolsyncpeer** - Specify IPs or subnets of peers with which we want to
periodically synchronise mempools. Can be used more than once to specify
multiple peers / networks.
*   **mempoolsyncage** - Only transactions older than this age will be included
in synchronisation.
*   **mempoolsyncperiod** - How often synchronisation takes place.

#### Other New Options

*   **preferredblockfilesize** - Preferred size (in bytes) of a single block datafile.
*   **maxcoinsdbfilesize** - Set maximum file size used by the coins database.
*   **maxconnections** - Maintain at most the given number of connections to peers.
*   **maxchroniclegracefulperiod** - The number of blocks either side of the Chronicle
activation height that we will give peers some grace if they are not yet fully
activated before banning them.

#### Removed Options

*   **maxsendbuffermult**

#### Changes to Existing Defaults

*   **dbcache** - The default database cache size is now 2GB.
*   **maxscriptnumlengthpolicy** - The default is now 0 (unlimited).

### Other items

*   STN Reset - includes an updated chain height block hash.

src/bitcoind.cpp
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2016 The Bitcoin Core developers
// Copyright (c) 2019 Bitcoin Association
// Distributed under the Open BSV software license, see the accompanying LICENSE.

#if defined(HAVE_CONFIG_H)
#include "config/bitcoin-config.h"
#endif

#include "chainparams.h"
#include "clientversion.h"
#include "compat.h"
#include "config.h"
#include "fs.h"
#include "init.h"
#include "noui.h"
#include "scheduler.h"
#include "util.h"
#include "utilstrencodings.h"
#include <boost/algorithm/string/predicate.hpp>
#include <boost/thread.hpp>

#include <cstdio>


/* Introduction text for doxygen: */

/*! \mainpage Developer documentation
 *
 * \section intro_sec Introduction
 *
 * This is the developer documentation of Bitcoin SV
 * (https://bitcoinsv.io/). Bitcoin SV is a client for the digital
 * currency called Bitcoin SV, which enables
 * instant payments to anyone, anywhere in the world. Bitcoin SV uses
 * peer-to-peer technology to operate with no central authority: managing
 * transactions and issuing money are carried out collectively by the network.
 *
 * The software is a community-driven open source project, released under the
 * Open BSV license.
 *
 * \section Navigation
 * Use the buttons <code>Namespaces</code>, <code>Classes</code> or
 * <code>Files</code> at the top of the page to start navigating the code.
 */

void WaitForShutdown(boost::thread_group *threadGroup, const task::CCancellationToken& shutdownToken) {

    // Tell the main threads to shutdown.
    while (!shutdownToken.IsCanceled()) {
        MilliSleep(200); // NOLINT(cppcoreguidelines-avoid-magic-numbers)
    }
    if (threadGroup) {
        Interrupt(*threadGroup);
        threadGroup->join_all();
    }
}

//////////////////////////////////////////////////////////////////////////////
//
// Start
//
// NOLINTBEGIN(cppcoreguidelines-pro-bounds-pointer-arithmetic)
// NOLINTBEGIN(cppcoreguidelines-pro-type-vararg)
// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
bool AppInit(int argc, char *argv[]) {
    RenameThread("main");
    boost::thread_group threadGroup;
    CScheduler scheduler;

    // FIXME: Ideally, we'd like to build the config here, but that's currently
    // not possible as the whole application has too many global state. However,
    // this is a first step.
    auto& config = GlobalConfig::GetModifiableGlobalConfig();

    bool fRet = false;

    //
    // Parameters
    //
    // main()
    gArgs.ParseParameters(argc, argv);

    // Process help and version before taking care about datadir
    if (gArgs.IsArgSet("-?") || gArgs.IsArgSet("-h") ||
        gArgs.IsArgSet("-help") || gArgs.IsArgSet("-version")) {
        std::string strUsage = strprintf(_("%s"), _(PACKAGE_NAME)) +
                               " " + _("version") + " " + FormatFullVersion() +
                               "\n";

        if (gArgs.IsArgSet("-version")) {
            strUsage += FormatParagraph(LicenseInfo());
        } else {
            strUsage += "\n" + _("Usage:") + "\n" +
                        "  bitcoind [options]                     " +
                        strprintf(_("Start %s Daemon"), _(PACKAGE_NAME)) + "\n";

            strUsage += "\n" + HelpMessage(HMM_BITCOIND, config);
        }

        std::cout << strUsage << '\n';
        return true;
    }

    try
    {
        if(!fs::is_directory(GetDataDir(false)))
        {
            std::ignore = fprintf(stderr,
                                  "Error: Specified data directory \"%s\" does not exist.\n",
                                  gArgs.GetArg("-datadir", "").c_str());
            return false;
        }

        try
        {
            gArgs.ReadConfigFile(gArgs.GetArg("-conf", BITCOIN_CONF_FILENAME));
        }
        catch(const std::exception& e)
        {
            std::cerr << "Error reading configuration file: " << e.what() << '\n';
            return false;
        }

        // Check for -testnet or -regtest parameter (Params() calls are only
        // valid after this clause)
        try
        {
            SelectParams(ChainNameFromCommandLine(), MagicBytesFromCommandLine());
        }
        catch(const std::exception& e)
        {
            std::cerr << "Error: " << e.what() << '\n';
            return false;
        }

        // Fill config with block size data
        config.SetDefaultBlockSizeParams(Params().GetDefaultBlockSizeParams());

        // maxstackmemoryusageconsensus and excessiveblocksize are required parameters
        if(!gArgs.IsArgSet("-maxstackmemoryusageconsensus") || !gArgs.IsArgSet("-excessiveblocksize"))
        {
            std::cerr << "Mandatory consensus parameter is not set. In order to "
                         "start bitcoind you must set the "
                         "following consensus parameters: "
                         "\"excessiveblocksize\" and "
                         "\"maxstackmemoryusageconsensus\". In order to start "
                         "bitcoind with no limits you can set "
                         "both of these parameters to 0 however it is strongly "
                         "recommended to ensure you understand "
                         "the implications of this setting.\n\n"
                         "For more information of how to choose these settings "
                         "safely for your use case refer to: "
                         "https://node.bitcoinsv.io/sv-node/installation/"
                         "sv-node/genesis-settings/";
            return false;
        }
        if (!gArgs.IsArgSet("-minminingtxfee"))
        {
            std::cerr << "Mandatory policy parameter is not set. In order to "
                         "start bitcoind you must set the "
                         "following policy parameters: \"minminingtxfee\"";
            return false;
        }

        // Command-line RPC
        bool fCommandLine = false;
        for (int i = 1; i < argc; i++)
            if (!IsSwitchChar(argv[i][0]) &&
                !boost::algorithm::istarts_with(argv[i], "bitcoin:"))
                fCommandLine = true;

        if(fCommandLine)
        {
            std::cerr << "Error: There is no RPC client functionality in "
                         "bitcoind anymore. Use the bitcoin-cli utility "
                         "instead.\n";
            exit(EXIT_FAILURE);
        }
        // -server defaults to true for bitcoind
        gArgs.SoftSetBoolArg("-server", true);
        // Set this early so that parameter interactions go to console
        InitLogging();
        InitParameterInteraction();
        if (!AppInitBasicSetup()) {
            // InitError will have been called with detailed error, which ends
            // up on console
            exit(1);
        }
        if (!AppInitParameterInteraction(config)) {
            // InitError will have been called with detailed error, which ends
            // up on console
            exit(1);
        }
        if (!AppInitSanityChecks()) {
            // InitError will have been called with detailed error, which ends
            // up on console
            exit(1);
        }
        if (gArgs.GetBoolArg("-daemon", false)) {
#if HAVE_DECL_DAEMON
            std::cout << "Bitcoin server starting\n";

            // Daemonize
            if (daemon(1, 0)) {
                // don't chdir (1), do close FDs (0)
                std::cerr << "Error: daemon() failed: " << strerror(errno) << '\n';
                return false;
            }
#else
            std::cerr << "Error: -daemon is not supported on this operating system\n";
            return false;
#endif // HAVE_DECL_DAEMON
        }

        fRet = AppInitMain(config, threadGroup, scheduler, GetShutdownToken());
    }
    catch(const std::exception& e)
    {
        PrintExceptionContinue(&e, "AppInit()");
    }
    catch(...)
    {
        PrintExceptionContinue(nullptr, "AppInit()");
    }
    GetAppInitCompleted().store(true);

    if (!fRet) {
        Interrupt(threadGroup);
        // threadGroup.join_all(); was left out intentionally here, because we
        // didn't re-test all of the startup-failure cases to make sure they
        // don't result in a hang due to some
        // thread-blocking-waiting-for-another-thread-during-startup case.
    } else {
        LogPrintf("Preload wait for shutdown\n");
        WaitForShutdown(&threadGroup, GetShutdownToken());
        LogPrintf("Preload wait for shutdown done\n");
    }
    LogPrintf("Checking Thread shutdown\n");
    Shutdown();

    return fRet;
}
// NOLINTEND(cppcoreguidelines-pro-type-vararg)
// NOLINTEND(cppcoreguidelines-pro-bounds-pointer-arithmetic)

int main(int argc, char *argv[]) { // NOLINT(bugprone-exception-escape)
    SetupEnvironment();

    // Connect bitcoind signal handlers
    noui_connect();

    return (AppInit(argc, argv) ? EXIT_SUCCESS : EXIT_FAILURE);
}
